src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1078,36 +996,340 @@ protected void workerDone(Worker worker) {
+          
+          scheduleQueue.add(index, task);
+        } finally {
+          clockWrapper.resumeForcingUpdate();
+        }
+      }
+      
+      handleQueueUpdate();
+    }
+
+    /**
+     * Adds a scheduled task to the end of the scheduled queue.  It is expected that this task is 
+     * NOT ready for execution, and will later be moved from invoking 
+     * {@link #reschedule(RecurringTaskWrapper)}.
+     * 
+     * @param task Task to add to end of schedule queue
+     */
+    public void addScheduledLast(RecurringTaskWrapper task) {
+      scheduleQueue.addLast(task);
+      // no need to notify since this task wont be ready to run
+    }
+
+    /**
+     * Call to find and reposition a scheduled task.  It is expected that the task provided has 
+     * already been added to the queue (likely from a call to 
+     * {@link #addScheduledLast(RecurringTaskWrapper)}).  This call will use 
+     * {@link RecurringTaskWrapper#getNextDelayInMillis()} to figure out what the new position 
+     * within the queue should be.
+     * 
+     * @param task Task to find in queue and reposition based off next delay
+     */
+    public void reschedule(RecurringTaskWrapper task) {
+      synchronized (scheduleQueue.getModificationLock()) {
+        if (! shutdownStarted.get()) {
+          clockWrapper.stopForcingUpdate();
+          try {
+            long nextDelay = task.getNextDelayInMillis();
+            int insertionIndex = ListUtils.getInsertionEndIndex(scheduleQueue, nextDelay, true);
+            
+            scheduleQueue.reposition(task, insertionIndex, true);
+          } finally {
+            clockWrapper.resumeForcingUpdate();
+          }
+        }
+      }
+      
+      // need to unpark even if the task is not ready, otherwise we may get stuck on an infinite park
+      handleQueueUpdate();
+    }
+
+    /**
+     * Called to check if either queue has anything to run.  This is just if the queues are empty.  
+     * If there are scheduled tasks queued, but not ready to run, this will still return 
+     * {@code false}.
+     * 
+     * @return {@code true} if there are no tasks in either queue
+     */
+    public boolean isQueueEmpty() {
+      return executeQueue.isEmpty() && scheduleQueue.isEmpty();
+    }
+
+    /**
+     * Call to get the total quantity of tasks within both stored queues.  If you can, 
+     * {@link #isQueueEmpty()} is a more efficient call.  This returns the total amount of items 
+     * in both the execute and scheduled queue.  If there are scheduled tasks which are NOT ready 
+     * to run, they will still be included in this total.
+     * 
+     * @return Total quantity of tasks queued
+     */
+    public int queueSize() {
+      return scheduleQueue.size() + executeQueue.size();
+    }
+
+    public void stopAndDrainQueueInto(List<Runnable> removedTasks) {
+      stopIfRunning();
+      
+      synchronized (executeQueueRemoveLock) {
```

Comment:
We have to lock on just the removes so we can atomically remove (ie if someone wants to remove from the .remove() invokation and at the same time trying to find which task to remove.

Your right, we don't need to lock on adds (though the scheduleQueue does lock on adds).